docs/en/integrations/build-integration.asciidoc

```diff
@@ -208,20 +208,55 @@ apache
 │   │   └───elasticsearch/ingest_pipeline
 │   │   │      default.yml
 │   │   └───fields
-│   │          agent.yml <1>
-│   │          base-fields.yml <2>
-│   │          ecs.yml <3>
-│   │          fields.yml <4>
+│   │          agent.yml
+│   │          base-fields.yml
+│   │          ecs.yml
+│   │          fields.yml
 │   └───error
 │   └───status
 ----
-<1> `agent.yml` fields the Elastic agent uses
-<2> `base-fields.yml` never changes and is required for all integrations
-<3> Defines the relevant ECS fields
-<4> Custom Apache access log fields
 
+=== agent.yml
+The `agent.yml` file defines fields used by default processors, this file usually remains unchanged between integrations.
+
+=== base-fields.yml
+The fields defined in this file are placeholders for the datastream subfields which, in turn, receive their value from the Elastic agent.
+In addition, the fields `event.module` and `event.dataset` are also present; the values in these fields depend on the integration.
+For example, in the Apache integration (more specifically, for the access dataset) that would be:
+* event.module: apache
+* event.dataset: apache.access
+
+=== ecs.yml
+This file specifies every ECS field used by the integration that is not defined in the files `agent.yml` or `base-fields.yml` files.
+To avoid redefining every ECS field in every integration, we can specify an existing field as external. For example:
+
+[source,yaml]
+----
+- external: ecs
+  name: client.ip
+- external: ecs
+  name: destination.domain
+----
+
+The `external: ecs` definition instructs the `elastic-package` binary to refer to an external ECS reference to resolve these fields. By default it looks at the ECS https://raw.githubusercontent.com/elastic/ecs/v8.5.1/generated/ecs/ecs_nested.yml[reference] file on Github, following the Git reference in the `_dev/build/build.yml` file located at the root of the package directory.
+By changing the reference to the ECS reference file in `_dev/build/build.yml` from `Git@..` to `file:///file/path/ecs_nested.yml`, it is possible to look for this file locally.
```

Comment:
Two things to add here. We support now two ways to import the whole ECS:
* Adding `import_mappings: true` into the `ecs` section in `_dev/build/build.yml` adds a dynamic mapping with most of the ECS definitions ([this one](https://github.com/elastic/elastic-package/blob/f439b96a74c27c5adfc3e7810ad584204bfaf85d/internal/builder/_static/ecs_mappings.yaml)).
* Starting on 8.13.0, Fleet pre-installs the [`ecs@mappings`](https://github.com/elastic/elasticsearch/blob/c2a3ec42632b0339387121efdef13f52c6c66848/x-pack/plugin/core/template-resources/src/main/resources/ecs%40mappings.json) component template into all integrations. This template also includes dynamic mappings for most of ECS fields.

In both cases, definitions in the fields files in packages take precedence, so normal definitions, or definitions with `external` can be used to override the dynamic templates.

Reply:
Good points!
It seems there is a lot more to explain in the other sections as well, bit a how-to doesn’t seem like the place for it. Is there some reference documentation where this can be explained in more length?
For some parts I had to dig through source code to be able to grok things to make my own integration. Would be nice if other people don’t have to.
Besides that, I really love the way integrations centralize everything.

Reply:
Yeah, sadly we have documentation spread along several places. You can find more docs at least in:
* https://github.com/elastic/elastic-package/tree/main/docs/howto
* https://github.com/elastic/integrations/tree/main/docs

+1 to centralize everything in a single place. Probably also taking into account development of inputs in Beats or other collectors.

Reply:
I'd be willing to contribute to the documentation (instead of having to write internal docs), but I'd need someone to discuss changes with in addition to a landing place for these centralized docs.

@bmorelli25 : What is your take on this?

Reply:
This is a topic that @nimarezainia, @kilfoyle, and I discussed recently. Based on Slack messages today, it sounds like we're all in favor of moving forward with the integrations developer guide (these docs) as our source of truth for all things developer-related.

@kilfoyle is going to be moving the new content in https://github.com/elastic/integrations/tree/main/docs to this guide. Beyond that, we're still formulating a plan.

We'd love any contribution you're willing to make to the docs. I'm happy to help organize and structure any content you want to contribute.

Is that helpful?

Reply:
Thanks for the clarification, that's definitely helpful.
So, to sum it up:
- I'll review @jsoriano above suggestions, make changes and (hopefully) get this MR merged.
- Eventually we can also put this information in the integration develop guide.
- I'll keep an eye on the integration develop docs to contribute.

Is there a semi real-time way we can use to communicate?

Reply:
That's mostly correct. The only clarification I'll add is:

> Eventually we can also put this information in the integration develop guide.

This PR will add this content to the integration developer guide. That's what we call the documentation set that you're adding this content too.

> Is there a semi real-time way we can use to communicate?

Hmmm. Unfortunately, not really. The only thing I can think of is the [Elastic Stack Community Slack](elasticstack.slack.com). My username in that Slack group is the same as it is on GitHub.

Reply:
> @jsoriano, can you confirm that in practice this means that ECS fields and types don't need to be specified in the integration, because they are dynamically picked up at build time (import_mappings option) or dynamically picked up in the component template in Elastic?

Correct. But only for packages with `import_mappings: true`, or that only support versions of the stack starting on 8.13.0.